### PR TITLE
feat: Sentry context

### PR DESCRIPTION
**Pull request by bruno-garcia**
* Payload creating has runtime detection
* Context is built lazily on first event
* Root context is closed per event to avoid sys calls required to build it
* .NET Framework 4.7.1 added to support NET Standard 2.0 paths (RuntimeInformation)
* Warnings over obsolete calls removed from test projects
* Nancy targets net471 also
* Json.NET indirect reference to Nancy

### CODE REVIEW

File: src/app/SharpRaven/Data/Context/App.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+                var app = new App();
+
+                var asm = ApplicationName ?? EntryAssemblyNameLocator.GetAssemblyName();
+                if (asm != null)
+                {
+                    app.Name = asm.Name;
+                    app.Version = asm.Version?.ToString();
+                }
+
+                using (var proc = Process.GetCurrentProcess())
```
Comment (asbjornu) [inline src/app/SharpRaven/Data/Context/App.cs]: Is it wise to `.Dispose()` the current process?

File: src/app/SharpRaven/Data/Context/Device.cs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+                SystemUtil.WriteError(e);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Retrieves the architecture of this device.
+        /// </summary>
+        /// <returns>Arch or null if unknown</returns>
+        internal static string GetArchitecture() =>
```
Comment (asbjornu) [inline src/app/SharpRaven/Data/Context/Device.cs]: All these `internal` methods only seem to be used in a `private` scope, so can they be changed to `private` instead?

File: src/app/SharpRaven/Data/SentryEvent.cs [hunk 3]
```diff
@@ -68,8 +70,15 @@ private SentryEvent()
...
             Tags = new Dictionary<string, string>();
             Fingerprint = new List<string>();
+            this.contexts = Contexts.Create();
         }
 
+        /// <summary>Represents Sentry's structured Context</summary>
+        /// <seealso href="https://docs.sentry.io/clientdev/interfaces/contexts/"/>
+        public Contexts Contexts
+        {
+            get { return this.contexts; }
```
Comment (asbjornu) [inline src/app/SharpRaven/Data/SentryEvent.cs]: Can't you just add an `internal set;` to this property so we get rid of the field?

File: src/app/SharpRaven/Utilities/EntryAssemblyNameLocator.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            return entryAssemblyName;
+        }
+
+        internal static AssemblyName GetAssemblyNameFromEntryMethod()
+        {
+            var entryMethod = GetApplicationEntryMethod();
+            return entryMethod?.Module?.Assembly?.GetName();
+        }
+
+        internal static MethodBase GetApplicationEntryMethod()
```
Comment (asbjornu) [inline src/app/SharpRaven/Utilities/EntryAssemblyNameLocator.cs]: 👍

File: src/app/SharpRaven/Utilities/RuntimeInfoHelper.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            string monoVersion = null;
+
+            var type = Type.GetType("Mono.Runtime", false);
+            if (type != null)
+            {
+                var displayName = type.GetMethod("GetDisplayName", BindingFlags.NonPublic | BindingFlags.Static);
+                if (displayName != null)
+                {
+                    monoVersion = displayName.Invoke(null, null) as string;
+                    if (monoVersion != null)
```
Comment (asbjornu) [inline src/app/SharpRaven/Utilities/RuntimeInfoHelper.cs]: Can't all of these `null` checks be collapsed into a chained `null` propagation (`?.`) expression?

File: src/app/SharpRaven/Utilities/RuntimeInfoHelper.cs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                        ? $".NET Framework {CheckFor45PlusVersion(releaseId)}"
+                        : null;
+                }
+            }
+            catch (PlatformNotSupportedException)
+            {
+                return null;
+            }
+
+            string CheckFor45PlusVersion(int releaseKey)
```
Comment (asbjornu) [inline src/app/SharpRaven/Utilities/RuntimeInfoHelper.cs]: A bit unintuitive that a method called `Check...()` returns `string`.

File: src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+namespace SharpRaven.UnitTests.Data.Context
+{
+    [TestFixture]
+    public class AppTests
+    {
+        [Test]
+        public void Create_StartTime_SometimeInThePast()
+        {
+            var sut = App.Create();
```
Comment (asbjornu) [inline src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs]: I'm not a fan of `sut`. Can't this just be called `app`?

File: src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+            Assert.True(sut.StartTime.HasValue);
```
Comment (asbjornu) [inline src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs]: I think `Assert.That(app.StartTime.HasValue, Is.True)` reads better.

File: src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            Assert.AreNotEqual(default(DateTimeOffset), sut.StartTime.Value);
```
Comment (asbjornu) [inline src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs]: I think `Assert.That(app.StartTime.Value, Is.Not.EqualTo(default(DateTimeOffset))` reads better.

File: src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+            Assert.Greater(DateTimeOffset.UtcNow, sut.StartTime);
```
Comment (asbjornu) [inline src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs]: I think `Assert.That(app.StartTime, Is.GreaterThan(DateTimeOffset.UtcNow))` reads better.

File: src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+#if !NET35
+        [Test]
+        public void Create_Version_EqualEntryAssembly()
+        {
+            var sut = App.Create();
+
+            Assert.NotNull(sut.Version);
```
Comment (asbjornu) [inline src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs]: I think `Assert.That(app.Version, Is.Not.Null)` reads better.

File: src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        [Theory]
+        [Test, TestCaseSource(typeof(AppTests), nameof(TestCases))]
+        public void SerializeObject_TestCase_SerializesAsExpected(TestCase @case)
+        {
+            var actual = JsonConvert.SerializeObject(@case.Object);
+
+            Assert.That(actual, Is.EqualTo(@case.ExpectedSerializationOutput));
+        }
+
+        public static IEnumerable<object[]> TestCases()
```
Comment (asbjornu) [inline src/tests/SharpRaven.UnitTests/Data/Context/AppTests.cs]: 👍

File: src/app/SharpRaven/Data/Context/App.cs (hunk 1 above)
Comment (bruno-garcia) [reply src/app/SharpRaven/Data/Context/App.cs]: Yes. This instance was created by me at this location and should be [disposed when not needed](https://stackoverflow.com/questions/28023490/should-one-call-dispose-for-process-getcurrentprocess).

File: src/app/SharpRaven/Data/Context/Device.cs (hunk 2 above)
Comment (bruno-garcia) [reply src/app/SharpRaven/Data/Context/Device.cs]: Yes, I made them internal for testing but I'm just covering them through testing the whole Device at the end. I'll make them private.

File: src/app/SharpRaven/Utilities/RuntimeInfoHelper.cs (hunk 5 above)
Comment (bruno-garcia) [reply src/app/SharpRaven/Utilities/RuntimeInfoHelper.cs]: I was afraid it'd be less readable but it's fine. Will change.

File: src/app/SharpRaven/Utilities/RuntimeInfoHelper.cs (hunk 6 above)
Comment (bruno-garcia) [reply src/app/SharpRaven/Utilities/RuntimeInfoHelper.cs]: True. I'll make it: Get

File: src/app/SharpRaven/Data/Context/Device.cs (hunk 2 above)
Comment (bruno-garcia) [reply src/app/SharpRaven/Data/Context/Device.cs]: It is actually used by tests*

**Comment by bruno-garcia** [comment]
@asbjornu could you please check again?

**Comment by asbjornu** [comment]
@bruno-garcia: `GetArchitecture` and `GetApplicationEntryMethod` are still `internal`, but otherwise I think it looks good.

**Comment by bruno-garcia** [comment]
@asbjornu I commented on the review: It's to be accessed by the Unit testing library.

**Comment by bruno-garcia** [comment]
@asbjornu if that's ok, could you please approve it? Thanks!

**Pull request merged**